examples/custom-libp2p/index.js

```diff
@@ -0,0 +1,129 @@
+'use strict'
+
+const Libp2p = require('libp2p')
+const IPFS = require('ipfs')
+const TCP = require('libp2p-tcp')
+const MulticastDNS = require('libp2p-mdns')
+const WebSocketStar = require('libp2p-websocket-star')
+const Bootstrap = require('libp2p-railing')
+const SPDY = require('libp2p-spdy')
+const KadDHT = require('libp2p-kad-dht')
+const MPLEX = require('libp2p-mplex')
+const SECIO = require('libp2p-secio')
+const assert = require('assert')
+
+/**
+ * Options for the libp2p factory
+ * @typedef {Object} libp2pFactory~options
+ * @property {PeerInfo} peerInfo - The PeerInfo of the IPFS node
+ * @property {PeerBook} peerBook - The PeerBook of the IPFS node
+ * @property {Object} config - The config of the IPFS node
+ * @property {Object} options - The options given to the IPFS node
+ */
+
+/**
+ * This is the factory we will use to create our fully customized libp2p node.
+ *
+ * @param {libp2pFactory~options} opts The options to use when generating the libp2p node
+ * @returns {Libp2p} Our new libp2p node
+ */
+const libp2pFactory = (opts) => {
+  // Set convenience variables to clearly showcase some of the useful things that are available
+  const peerInfo = opts.peerInfo
+  const peerBook = opts.peerBook
+  const bootstrapList = opts.config.Bootstrap
+
+  // Create our WebSocketStar transport and give it our PeerId, straight from the ipfs node
+  const wsstar = new WebSocketStar({
+    id: peerInfo.id
+  })
+
+  // Build and return our libp2p node
+  return new Libp2p({
+    peerInfo,
+    peerBook,
+    // Lets limit the connection managers peers and have it check peer health less frequently
+    connectionManager: {
+      maxPeers: 25,
+      pollInterval: 5000
+    },
+    modules: {
+      transport: [
+        TCP,
+        wsstar
+      ],
+      streamMuxer: [
+        MPLEX,
+        SPDY
+      ],
+      connEncryption: [
+        SECIO
+      ],
+      peerDiscovery: [
+        MulticastDNS,
+        Bootstrap,
+        wsstar.discovery
+      ],
+      dht: KadDHT
+    },
+    config: {
+      peerDiscovery: {
+        mdns: {
+          interval: 10000,
+          enabled: true
+        },
+        bootstrap: {
+          interval: 10000,
+          enabled: true,
+          list: bootstrapList
+        }
+      },
+      // Turn on relay with hop active so we can connect to more peers
+      relay: {
+        enabled: true,
+        hop: {
+          enabled: true,
+          active: true
+        }
+      },
+      dht: {
+        kBucketSize: 20
+      },
+      EXPERIMENTAL: {
+        dht: true,
+        pubsub: true
+      }
+    }
+  })
+}
+
+// Now that we have our custom factory, let's start up the ipfs node!
+const node = new IPFS({
+  libp2p: libp2pFactory
+})
```

Comment:
Instead of calling it Factory, let's just reuse the wording around libp2p bundles (used in https://github.com/libp2p/js-libp2p#creating-your-own-libp2p-bundle, the website, slidedecks and other places).

Also, it would be great that instead of passing a factory, we could pass the libp2p instance itself.

Reply:
I've updated the language to be more bundle centric, and added a link to the example in the main examples readme since it wasn't there yet.